src/main/scala/ai/privado/languageEngine/go/tagger/collection/CollectionTagger.scala

```diff
@@ -0,0 +1,173 @@
+  }
+
+  private def tagRestCallCollection(builder: DiffGraphBuilder, sourceRuleInfo: RuleInfo): Unit = {
+
+    val collectionCallMethod = cpg
+      .call(COLLECTION_REST_PATTERN)
+      .where(_.file.name(ROUTES_FILE_PATTERN))
+      .l
+
+    // sample route -> r.Get("/user/{id}", getUser)
+    val collectionMethodsCache = collectionCallMethod
+      .map { m =>
+        if (m.argument.size >= 2) {
+          val targetCollectionUrl =
+            m.argument(1).code
+
+          if (targetCollectionUrl.nonEmpty) {
+            val methodName             = m.argument(2).code
+            val targetCollectionMethod = cpg.method.name(methodName).l
+            if (targetCollectionMethod.nonEmpty) {
+              methodUrlMap.addOne(targetCollectionMethod.head.id -> m.argument(1).code)
+              targetCollectionMethod
+            } else None
+          } else None
+        } else None
+      }
+      .l
+      .flatten(method => method)
+
+    tagDirectSource(cpg, builder, collectionMethodsCache.l, sourceRuleInfo)
+  }
+
+  private def tagDirectSource(
+    cpg: Cpg,
+    builder: DiffGraphBuilder,
+    collectionMethods: List[Method],
+    sourceRuleInfo: RuleInfo
+  ): Unit = {
+    val collectionPoints = collectionMethods.flatMap(collectionMethod => {
+      val parameters = collectionMethod.parameter
+      val locals     = collectionMethod.local
+      val literals   = collectionMethod.literal
+
+      val matchingParameters = parameters.where(_.name(sourceRuleInfo.combinedRulePattern)).whereNot(_.code("self")).l
+      val matchingLocals     = locals.code(sourceRuleInfo.combinedRulePattern).l
+      val matchingLiterals = literals
+        .code(sourceRuleInfo.combinedRulePattern)
+        .l
+
+      if (!(matchingParameters.isEmpty && matchingLocals.isEmpty && matchingLiterals.isEmpty)) {
+        matchingParameters.foreach(parameter =>
+          storeForTag(builder, parameter, ruleCache)(Constants.id, sourceRuleInfo.id)
+        )
+        matchingLocals.foreach(local => storeForTag(builder, local, ruleCache)(Constants.id, sourceRuleInfo.id))
+        matchingLiterals.foreach(literal => storeForTag(builder, literal, ruleCache)(Constants.id, sourceRuleInfo.id))
+        Some(collectionMethod)
+      } else {
+        None
+      }
+    })
+
+    tagMethodEndpoints(builder, collectionPoints.l, sourceRuleInfo)
+  }
+
+  private def tagMethodEndpoints(
+    builder: DiffGraphBuilder,
+    collectionPoints: List[Method],
+    sourceRuleInfo: RuleInfo,
+    returnByName: Boolean = false
+  ) = {
+    collectionPoints.foreach(collectionPoint => {
+      addRuleTags(builder, collectionPoint, getCollectionRule, ruleCache)
+      storeForTag(builder, collectionPoint, ruleCache)(
+        InternalTag.COLLECTION_METHOD_ENDPOINT.toString,
+        getFinalEndPoint(collectionPoint, returnByName)
+      )
+    })
+  }
+
+  /*
+  collection Rule used for method tagging only
+   */
+  private def getCollectionRule: RuleInfo = {
```

Comment:
can we define it at the top or create a separate file.

Reply:
done.